src/main/java/seedu/address/logic/UndoRedoStack.java

```diff
@@ -0,0 +1,96 @@
+package seedu.address.logic;
+
+import java.util.LinkedList;
+
+import seedu.address.logic.commands.Command;
+import seedu.address.logic.commands.RedoCommand;
+import seedu.address.logic.commands.UndoCommand;
+import seedu.address.logic.commands.UndoableCommand;
+
+/**
+ * Undo-redo Stack
+ * Solution adapted from https://github.com/se-edu/addressbook-level4/pull/610/files with modifications / bug fixes
+ */
+public class UndoRedoStack {
+
+    private LinkedList<UndoableCommand> undoStack;
+    private LinkedList<UndoableCommand> redoStack;
+
+    public UndoRedoStack() {
+        undoStack = new LinkedList<>();
+        redoStack = new LinkedList<>();
+    }
+
+    public UndoRedoStack(UndoRedoStack otherUndoRedoStack) {
+        undoStack = otherUndoRedoStack.undoStack;
+        redoStack = otherUndoRedoStack.redoStack;
+    }
+
+    /**
+     * Pushes {@code command} onto the undo-stack if it is of type {@code UndoableCommand}. Clears the redo-stack
+     * if {@code command} is not of type {@code UndoCommand} or {@code RedoCommand}.
+     */
+    public void push(Command command) {
+        if (!(command instanceof UndoableCommand)) {
+            return;
+        }
+
+        if (!(command instanceof UndoCommand) && !(command instanceof RedoCommand)) {
+            redoStack.clear();
+        }
+
+
+        undoStack.addFirst((UndoableCommand) command);
```

Comment:
You could just use `pop` here as in the [documentation](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/LinkedList.html) for `LinkedList`